core/src/main/java/com/datastrato/gravitino/authorization/PermissionManager.java

```diff
@@ -0,0 +1,319 @@
+    try {
+      AuthorizationUtils.checkMetalakeExists(store, metalake);
+
+      checkRoleExists(metalake, role);
+
+      RoleEntity roleEntity =
+          store.get(
+              NameIdentifierUtils.ofRole(metalake, role), Entity.EntityType.ROLE, RoleEntity.class);
+
+      AtomicBoolean removed = new AtomicBoolean(true);
+
+      store.update(
+          NameIdentifierUtils.ofUser(metalake, user),
+          UserEntity.class,
+          Entity.EntityType.USER,
+          userEntity -> {
+            List<RoleEntity> roleEntities = removeInvalidRoles(metalake, userEntity.roles());
+            List<String> roleNames = Lists.newArrayList(toRoleNames(roleEntities));
+            List<Long> roleIds = Lists.newArrayList(toRoleIds(roleEntities));
+            roleNames.remove(roleEntity.name());
+            removed.set(roleIds.remove(roleEntity.id()));
+            if (!removed.get()) {
+              LOG.warn("Role {} doesn't exist in the user {} of metalake {}", role, user, metalake);
+            }
+
+            AuditInfo auditInfo =
+                AuditInfo.builder()
+                    .withCreator(userEntity.auditInfo().creator())
+                    .withCreateTime(userEntity.auditInfo().createTime())
+                    .withLastModifier(PrincipalUtils.getCurrentPrincipal().getName())
+                    .withLastModifiedTime(Instant.now())
+                    .build();
+            return UserEntity.builder()
+                .withId(userEntity.id())
+                .withNamespace(userEntity.namespace())
+                .withName(userEntity.name())
+                .withRoleNames(roleNames)
+                .withRoleIds(roleIds)
+                .withAuditInfo(auditInfo)
+                .build();
+          });
+      return removed.get();
+    } catch (NoSuchEntityException nse) {
+      LOG.warn("User {} does not exist in the metalake {}", user, metalake, nse);
+      throw new NoSuchUserException(USER_DOES_NOT_EXIST_MSG, user, metalake);
+    } catch (IOException ioe) {
+      LOG.error(
+          "Revoking role {} from  user {} failed in the metalake {} due to storage issues",
+          role,
+          user,
+          metalake,
+          ioe);
+      throw new RuntimeException(ioe);
+    }
+  }
+
+  private List<String> toRoleNames(List<RoleEntity> roleEntities) {
+    return roleEntities.stream().map(RoleEntity::name).collect(Collectors.toList());
+  }
+
+  private List<Long> toRoleIds(List<RoleEntity> roleEntities) {
+    return roleEntities.stream().map(RoleEntity::id).collect(Collectors.toList());
+  }
+
+  private void checkRoleExists(String metalake, String role) throws IOException {
+    if (!store.exists(NameIdentifierUtils.ofRole(metalake, role), Entity.EntityType.ROLE)) {
+      throw new NoSuchRoleException(AuthorizationUtils.ROLE_DOES_NOT_EXIST_MSG, role, metalake);
+    }
+  }
+
+  private List<RoleEntity> removeInvalidRoles(String metalake, List<String> roles) {
```

Comment:
It has a high cost to do this, do we have a better solution?

Reply:
Maybe, this is a trade off:
Because we don't have transaction interface, we can' guarantee that we can remove role from users if we drop the role. 
If we can tolerance the consistency issue. 
We can remove role form users first. Then we drop the role.

Reply:
Can we have different solutions, like async check or doing this lazily, or do it when use, I'm wondering if there's a better solution to avoid heavy point looks here.

Reply:
1. The challenge of async check is the issue of the lock. It's a little hard.
2. I can remove the method `removeInvalidRoles` in the process of writing. But the situation which we faced is that more read and less write. So I add the `removeInvalidRoles` in the process of writing to improve the performance of read process.

Reply:
Another solution.
Our situation is that we drop role infrequently. 
So we can record the latest timestamp of dropping roles in the server memory and record the latest timestamp of `removeInvalidRoles` in the User/Group entity.  We can avoid using this in the most moment.

Reply:
> 1. The challenge of async check is the issue of the lock. It's a little hard.
> 2. I can remove the method `removeInvalidRoles` in the process of writing. But the situation which we faced is that more read and less write. So I add the `removeInvalidRoles` in the process of writing to improve the performance of read process.

I think you still cannot guarantee that all the roles are valid when you use it. Because the user can delete the role after writing, am I right?

Reply:
In the first solution, we should rewrite the user and group entity after the role is deleted.


> Another solution. Our situation is that we drop role infrequently. So we can record the latest timestamp of dropping roles in the server memory and record the latest timestamp of `removeInvalidRoles` in the User/Group entity. We can avoid using this in the most moment. @jerryshao WDYT? Is it ok?

We can compare the entity update timestamp and role drop timestamp to decide whether to use `removeInvalidRoles`.

Reply:
From your code, you did this `removeInvalidRoles` everywhere you use it, it's not like what you mentioned only did this here.

Reply:
I used removeInvalidRoles in the UserGroupManager for getUser and getGroup and this class.